gravitee-am-resource/gravitee-am-resource-infobip/src/main/java/io/gravitee/am/resource/infobip/provider/InfobipResourceProvider.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.resource.infobip.provider;
+
+import com.infobip.ApiClient;
+import com.infobip.api.TfaApi;
+import com.infobip.model.TfaStartAuthenticationRequest;
+import com.infobip.model.TfaStartAuthenticationResponse;
+import com.infobip.model.TfaVerifyPinRequest;
+import com.infobip.model.TfaVerifyPinResponse;
+import io.gravitee.am.common.exception.mfa.InvalidCodeException;
+import io.gravitee.am.common.exception.mfa.SendChallengeException;
+import io.gravitee.am.resource.api.mfa.MFAChallenge;
+import io.gravitee.am.resource.api.mfa.MFALink;
+import io.gravitee.am.resource.api.mfa.MFAResourceProvider;
+import io.gravitee.am.resource.infobip.InfobipResourceConfiguration;
+import io.reactivex.Completable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * @author Ruan Ferreira (ruan@incentive.me)
+ * @author Incentive.me
+ */
+public class InfobipResourceProvider implements MFAResourceProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InfobipResourceProvider.class);
+
+    TfaApi tfaApi;
+    public static final String APPROVED = "approved";
+
+    private String pinId;
+    private String to;
+
+    @Autowired
+    private InfobipResourceConfiguration configuration;
+
+    @Override
+    public Completable send(MFALink target) {
+        this.to = target.getTarget();
+
+        String baseUrl = this.configuration.getBaseUrl();
+        String apiKey = this.configuration.getApiKey();
+        String apiKeyPrefix = this.configuration.getApiKeyPrefix();
+        String messageId = this.configuration.getMessageId();
+        String applicationId = this.configuration.getApplicationId();
+
+        ApiClient apiClient = new ApiClient();
+        apiClient.setApiKeyPrefix(apiKeyPrefix);
+        apiClient.setApiKey(apiKey);
+        apiClient.setBasePath(baseUrl);
+
+        this.tfaApi = new TfaApi(apiClient);
```

Comment:
Client creation should be done only one time at start up (afterPropertiesSet() method for example). This way you avoid to initiate the client every request.

Reply:
Hi @tcompiegne
So, I see on the twilio plugin, they use this:
```java
 @Override
    public ResourceProvider start() throws Exception {
        Twilio.init(configuration.getAccountSid(), configuration.getAuthToken());
        return this;
    }

    @Override
    public ResourceProvider stop() throws Exception {
        Twilio.destroy();
        return this;
    }
```

Do I need to put the client instantiation on the `start` method?

Reply:
Yes exactly, and if there is a destroy method for your client, you can call it in the stop() method.

Reply:
I changed it, but just for curiosity: what would be the `afterPropertiesSet()`? I looked into the interfaces of `MFAResourceProvider` and didn't find it. Or is it more like that I'm setting the configurations and then, **after it**, I do:
```java
this.tfaApi = new TfaApi(apiClient);
```

Is just I'm confused if `afterPropertiesSet()` is a method hahahaha

Reply:
:) It's just the method of InitializingBean Spring object. Which can also be called in addition of the start method. But start method is fine.